### DISCUSSION TITLE
High energy consumption when using machine.lightsleep

### DISCUSSION THREAD

**Discussion by mu88** [micropython/micropython#12882] [RP2040 / Pico] [answered]
Dear community,

As my first MicroPython project, I've connected a Waveshare eInk display with a RP2040 and display some temperature data ([see here](https://github.com/mu88/ScreenshotCreator)). It works nicely, so I started digging deeper and bought an USB stick to measure the Raspi's power consumption, which confuses me a little bit.

[Here](https://github.com/mu88/ScreenshotCreator/blob/main/src/WaveshareDisplayCode/main.py) you can find the complete MicroPython code, but the process looks like this:
1. Connect to WiFi
2. Pull new temperature data (via `HTTP GET`, returns binary data)
3. Update screen with new temperature data
4. Sleep for 5 min via `machine.lightsleep(300000)`

This code runs in a `while` loop.

What makes me wonder is that after calling `machine.lightsleep`, the power consumption is still between 48 and 52 mA. Why is this the case? Do I have to manually turn off the WiFi before calling `machine.lightsleep`? Or should I forget about the `while` loop and use `machine.deepsleep(300000)` so that my MicroPython script runs from the very beginning every 5 min?

**Comment by davefes** [comment, marked as answer]
The only reason, that I can think of, for using lightsleep on the ESP32 is that you can retain program state, ie pin hold.  It is like a pause.  Yes, you need to do a `sta.active(False)`before going to lightsleep and then a `sta.active(True)` when waking-up.

If you are not wanting to hold any output ports then deepsleep could save you more current.  Be aware, if you are using a dev board with a USB to UART chip on it that it sucks current as well.

**Comment by robert-hh** [comment]
I found an interesting behavior with the Pico W. If I just start it and call machine.lightsleep, the current drops to ~13mA. If I call
```
import network
wlan=network.WLAN(network.STA_IF)
wlan.deinit()
machine.lightsleep(time)
```
the current drops to < 1mA. That even valid when WiFi is not used otherwise. With wlan.deinit() the LED turns off, which alone contributes to ~5 mA current.

**Reply by davefes** [reply]
Right, this is a much better answer!  I had noticed that behaviour on the Pico W as well, but somehow tried to relate it to my ESP32 experience. 

Does lightsleep on Pico W allow you to hold pins and/or retain program state?

**Reply by robert-hh** [reply]
The pin state should be retained, and the code state as well. Only the clock freq. is reduced and USB switched off.

**Reply by robert-hh** [reply]
Supplying negative values to lightsleep(n)/deepsleep(n) will sleep almost forever. It then perform to code path of time.sleep_ms(n).

**Reply by davefes** [reply]
Thank you for the clarification.
